File: Fas7ny BackEnd/app/api/routes/planner.py
import json
import logging

from fastapi import APIRouter, HTTPException
from google.genai.errors import ServerError, ClientError
from pydantic import ValidationError

from app.models.request import TripRequest
from app.services.planner_service import generate_trip

logger = logging.getLogger("planner_routes")

router = APIRouter(
    prefix="/planner",
    tags=["Trip Planner"]
)

# كل الأنواع دي معناها فعليًا "كل مصادر الـ AI فشلت" - مش بس Gemini
# (ServerError/ClientError)، لكن كمان لما Groq نفسه يكون آخر محاولة
# وفشل بإرجاع JSON غير صالح (json.JSONDecodeError) أو JSON صالح لكن
# مش مطابق لـ TripPlan schema (pydantic ValidationError). اتأكدنا من
# ده بمراجعة _generate_with_groq() في gemini_service.py مباشرة.
AI_UNAVAILABLE_ERRORS = (ServerError, ClientError, json.JSONDecodeError, ValidationError)


@router.post("/generate")
async def generate(request: TripRequest):
    """
    بيولّد خطة رحلة كاملة. الأخطاء الجزئية (فشل جلب صورة/rating لمكان
    معين، أو فشل الطقس ليوم معين) متعالجة بالفعل جوه enrich_trip()
    ومبتوصلش هنا - الرحلة بترجع كاملة مع بعض الحقول الفارغة بس.

    اللي ممكن يوصل هنا فعليًا هو فشل توليد الخطة بالكامل. ده بيحصل في
    حالتين حسب مراجعة gemini_service.py:
      1) كل مفاتيح Gemini فشلوا بـ ServerError/ClientError، وGroq
         نفسه رمى نفس النوعين (نادر لكن ممكن لو في مشكلة شبكة).
      2) كل مفاتيح Gemini فشلوا، وGroq رجع رد لكن مش JSON صالح أو مش
         مطابق لـ TripPlan schema (json.JSONDecodeError أو
         pydantic.ValidationError من _generate_with_groq).

    في الحالتين دول منسربش تفاصيل تقنية داخلية (اسم الـ exception،
    stack trace، أي جزء من الـ API keys) للمستخدم - بنرجع رسالة عربية
    واضحة بس، ونسجل التفاصيل الكاملة في اللوج عشان التشخيص.
    """
    try:
        trip = generate_trip(request.model_dump(mode="json"))
        return trip

    except AI_UNAVAILABLE_ERRORS as e:
        logger.error(
            "All AI providers failed for trip generation | destinations=%s | error_type=%s | error=%s",
            request.destinations, type(e).__name__, str(e),
        )
        raise HTTPException(
            status_code=503,
            detail="الخدمة مشغولة جدًا حاليًا، برجاء المحاولة تاني بعد لحظات.",
        )

    except Exception as e:
        # يشمل حالة "raise Exception('Failed to generate trip.')" العامة
        # اللي بترجع من planner_service.py لو trip رجع None بطريقة ما،
        # وأي حاجة تانية غير متوقعة تمامًا.
        logger.exception(
            "Unexpected error during trip generation | destinations=%s",
            request.destinations,
        )
        raise HTTPException(
            status_code=500,
            detail="حدث خطأ غير متوقع أثناء إنشاء الرحلة، برجاء المحاولة تاني.",
        )

File: Fas7ny Tourist/src/services/planner.js
const API = "http://127.0.0.1:8000";

/**
 * خطأ مخصص بيحمل معلومات كافية للفرونت يقرر يعرض إيه للمستخدم:
 * - status: كود الـ HTTP (503 = مشغول مؤقتًا، 500 = خطأ غير متوقع، 422 = بيانات غير صالحة)
 * - detail: الرسالة العربية الجاهزة من الباك اند (planner_routes.py بيرجعها دايمًا نص واضح، مش تفاصيل تقنية)
 * - isRetryable: هل منطقي نقترح على المستخدم يجرب تاني فورًا
 */
export class TripGenerationError extends Error {
  constructor(status, detail) {
    super(detail || "حدث خطأ أثناء إنشاء الرحلة")
    this.name = "TripGenerationError"
    this.status = status
    this.detail = detail
    // 503 = الخدمة مشغولة مؤقتًا (كل مصادر الـ AI فشلت) - يستاهل تجربة تانية
    // 422 = بيانات الطلب نفسها فيها مشكلة - إعادة المحاولة بنفس البيانات مش هتفيد
    // 500 / أي حاجة تانية = خطأ غير متوقع
    this.isRetryable = status === 503
  }
}

export async function generateTrip(data) {
  let response
  try {
    response = await fetch(`${API}/planner/generate`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    })
  } catch (networkError) {
    // فشل الـ fetch نفسه (السيرفر مش شغال، مفيش نت، CORS...) - مش رد
    // من الباك اند خالص، فبنفرّقه عن أخطاء الـ HTTP العادية
    throw new TripGenerationError(
      0,
      "تعذر الاتصال بالسيرفر. تأكد إن اتصالك بالإنترنت شغال وحاول تاني."
    )
  }

  if (!response.ok) {
    let detail = "حدث خطأ أثناء إنشاء الرحلة. حاول مرة أخرى."
    try {
      const body = await response.json()
      // FastAPI's HTTPException بيرجع { "detail": "..." } - ده الشكل
      // اللي planner_routes.py بيضمنه دايمًا (رسالة عربية جاهزة،
      // مفيش أي تفاصيل تقنية داخلية)
      if (body?.detail) detail = body.detail
    } catch {
      // الرد مش JSON أصلاً (نادر، ممكن لو حصل خطأ قبل FastAPI نفسه
      // زي proxy/gateway error) - نسيب الرسالة الافتراضية
    }
    throw new TripGenerationError(response.status, detail)
  }

  return await response.json()
}